tools/garaga_rs/src/poseidon_transcript.rs

```diff
@@ -0,0 +1,167 @@
+use crate::io::parse_field_elements_from_list;
+use lambdaworks_crypto::hash::poseidon::{starknet::PoseidonCairoStark252, Poseidon};
+use lambdaworks_math::field::{
+    element::FieldElement, fields::fft_friendly::stark_252_prime_field::Stark252PrimeField,
+};
+use num_bigint::BigUint;
+
+pub struct CairoPoseidonTranscript {
+    pub init_hash: FieldElement<Stark252PrimeField>,
+    pub state: [FieldElement<Stark252PrimeField>; 3],
+}
+
+pub fn bigint_split_4_96(x: &BigUint) -> [FieldElement<Stark252PrimeField>; 4] {
+    let one = &BigUint::from(1usize);
+    assert!(x < &(one << 384));
+    let base = one << 96;
+    let mask = &(base - one);
+    let limbs = [x & mask, (x >> 96) & mask, (x >> 192) & mask, x >> 288];
+    let elems = parse_field_elements_from_list::<Stark252PrimeField>(&limbs).unwrap();
+    [elems[0], elems[1], elems[2], elems[3]]
+}
+
+pub fn bigint_split_2_128(x: &BigUint) -> [FieldElement<Stark252PrimeField>; 2] {
+    let one = &BigUint::from(1usize);
+    assert!(x < &(one << 256));
+    let base = one << 128;
+    let mask = &(base - one);
+    let limbs = [x & mask, x >> 128];
+    let elems = parse_field_elements_from_list::<Stark252PrimeField>(&limbs).unwrap();
+    [elems[0], elems[1]]
+}
+
+impl CairoPoseidonTranscript {
+    pub fn new(init_hash: FieldElement<Stark252PrimeField>) -> Self {
+        let mut state = [init_hash, FieldElement::zero(), FieldElement::one()];
+        PoseidonCairoStark252::hades_permutation(&mut state);
+        Self { init_hash, state }
+    }
+
+    pub fn continuable_hash(&self) -> FieldElement<Stark252PrimeField> {
+        self.state[0]
+    }
+
+    pub fn rlc_coeff(&mut self) -> FieldElement<Stark252PrimeField> {
+        self.state[1]
+    }
+
+    pub fn update_sponge_state(
+        &mut self,
+        x: FieldElement<Stark252PrimeField>,
+        y: FieldElement<Stark252PrimeField>,
+    ) {
+        self.state[0] += x;
+        self.state[1] += y;
+        PoseidonCairoStark252::hades_permutation(&mut self.state);
+    }
+
+    pub fn hash_element(
+        &mut self,
+        x: &BigUint,
+    ) -> (
+        FieldElement<Stark252PrimeField>,
+        FieldElement<Stark252PrimeField>,
+    ) {
+        let base = FieldElement::<Stark252PrimeField>::one()
+            .double()
+            .pow(96usize);
```

Comment:
Create a constant BASE_96_FELT252 to avoid computing 2^96 mod p at each invocation